CONTRIBUTING.md

```diff
@@ -62,6 +62,10 @@ The project uses [GoogleStyle](https://google.github.io/styleguide/javaguide.htm
 You can install this code style into your IDE to make things more automatic:
  * [IntelliJ code style xml file](https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml)
  * [Eclipse code style xml file](https://github.com/google/styleguide/blob/gh-pages/eclipse-java-google-style.xml)
+ 
+In addition the project has the following style rules over and above the standard Google rules:
+
+**Final fields, parameters and local variables**
```

Comment:
Personally I am not a fan of making params and local vars final.

I used this style myself in previous projects of mine (HornetQ / ActiveMQ Artemis) because I drunk the koolaid and thought it would help prevent bugs. In later projects (e.g. Vert.x) we didn't do this, and we didn't notice any spike in bugs. Most probably because these bugs are rare in the first place.

Imho for any chance which is so invasive and adds such a significant amount of visual clutter I think there has to be an extremely strong justification for it, and in this case the justification is just not there. I think with a lot of defensive actions like this, they are defending against an imaginary foe!

We should also note that most projects don't have this restriction. Standard google codestyle doesn't include it (which is also the Confluent company standard). Kafka core doesn't have this restriction either.

I'd argue we should drop this rule.

Reply:
I don't really care about parameters/local variables (they're all transient anyway) but I feel that explicitly labeling class fields that are final with `final` adds value to the readability of the code. It's nice to know when something isn't supposed to change (especially primitives and non-container objects) and I find the clustering of final fields together at the top of the class before the non-final fields is nice.

Reply:
I agree, final class fields are a good idea. Just the params and local vars I don't think add value.

Reply:
Personally, I find having `final` parameters means I don't need to check the code to see if the variable is changed as I know its not. Likewise, if a param is not `final` then I know it _is_ changed, otherwise the checkstyle rule would object.

I have spent time tracking down bugs that turn out to be caused by people making changes and not realising that a var/param is mutable. However, this is not the main reason for having `final` vars in my opinion. The main benefit is it makes the code easier to grok.

Out checkstyle rules already pick this up and IDEs can be configured to auto-insert `final` where needed. So I see the overhead of this rule as very minimal. I don't see any benefit from dropping the rule IMHO.

Reply:
Final in fields/methods/classes adds value. Final in parameters/vars is arguable. I haven't found one bug due to not doing that in nearly 2 decades.

Reply:
> So I see the overhead of this rule as very minimal.

I don't think we should underestimate the negative effects of unnecessary verbosity in the code. Java is already very verbose, and any extra verbosity takes mental energy to filter out (whether conscious or not) which results in lower productivity.

There seems little if any evidence that this pattern helps catch significant amounts of bugs, and the argument on being able to see if a variable is changed seems weak (just highlight the variable name in IntelliJ and it will show you everywhere in the method where it is accessed).

For a change that has such a sweeping effect on code verbosity and readability (negative) I think the argument for having it has to be extremely strong. And I think that argument just isn't there.

Reply:
> Personally, I find having final params / vars means I don't need to check the code to see if the variable is changed as I know its not. Likewise, if a param is not final then I know it is changed, otherwise the checkstyle rule would object.

This is the main benefit to me as well, and why my vote is to keep this rule. It adds less value if we stick to smaller, simpler functions. Agree that it's not super useful for catching bugs though.

Reply:
FYI - IntelliJ has a feature that lets you highlight (I personally choose underscore) any reassigned param/local variable. I find this even more useful than labeling things final because I don't even need to look up at the declaration :P